gtsam/slam/PlanarProjectionFactor.h

```diff
@@ -0,0 +1,318 @@
+                Pose3 camera_pose = offset_pose.compose(CAM_COORD, H00);
+                PinholeCamera<Cal3DS2> camera = PinholeCamera<Cal3DS2>(camera_pose, calib);
+                if (H2 || H3) {
+                    // Dpose is for pose3, 2x6 (R,t)
+                    gtsam::Matrix Dpose;
+                    Point2 result = camera.project(landmark, Dpose, H1, H4);
+                    gtsam::Matrix DposeOffset = Dpose * H00; // 2x6
+                    if (H3)
+                        *H3 = DposeOffset; // with Eigen this is a deep copy (!)
+                    if (H2) {
+                        gtsam::Matrix DposeOffsetFwd = DposeOffset * H0;
+                        *H2 = Matrix::Zero(2, 3);
+                        (*H2)(0, 0) = DposeOffsetFwd(0, 3); // du/dx
+                        (*H2)(1, 0) = DposeOffsetFwd(1, 3); // dv/dx
+                        (*H2)(0, 1) = DposeOffsetFwd(0, 4); // du/dy
+                        (*H2)(1, 1) = DposeOffsetFwd(1, 4); // dv/dy
+                        (*H2)(0, 2) = DposeOffsetFwd(0, 2); // du/dyaw
+                        (*H2)(1, 2) = DposeOffsetFwd(1, 2); // dv/dyaw
+                    }
+                    return result;
+                } else {
+                    return camera.project(landmark, {}, {}, {});
+                }
+            } catch (CheiralityException& e) {
+                std::cout << "****** CHIRALITY EXCEPTION ******\n";
+                if (H1) *H1 = Matrix::Zero(2, 3);
+                if (H2) *H2 = Matrix::Zero(2, 3);
+                if (H3) *H3 = Matrix::Zero(2, 6);
+                if (H4) *H4 = Matrix::Zero(2, 9);
+                // return a large error
+                return Matrix::Constant(2, 1, 2.0 * calib.fx());
+            }
+        }
+
+        Point2 measured_; // pixel measurement
+
+    private:
+        static const Pose3 CAM_COORD;
+    };
+
+    // camera "zero" is facing +z; this turns it to face +x
+    const Pose3 PlanarProjectionFactorBase::CAM_COORD = Pose3(
+        Rot3(0, 0, 1,//
+            -1, 0, 0, //
+            0, -1, 0),
+        Vector3(0, 0, 0)
+    );
+
+    /**
+     * @class PlanarProjectionFactor1
+     * @brief One variable: the pose.
+     * Landmark, camera offset, camera calibration are constant.
+     * This is intended for localization within a known map.
+     */
+    class PlanarProjectionFactor1
+        : public PlanarProjectionFactorBase, public NoiseModelFactorN<Pose2> {
+    public:
+        typedef NoiseModelFactorN<Pose2> Base;
+        using Base::evaluateError;
+        PlanarProjectionFactor1() {}
+
+        ~PlanarProjectionFactor1() override {}
+
+        /// @return a deep copy of this factor
+        gtsam::NonlinearFactor::shared_ptr clone() const override {
+            return std::static_pointer_cast<gtsam::NonlinearFactor>(
+                gtsam::NonlinearFactor::shared_ptr(new PlanarProjectionFactor1(*this)));
+        }
+
+
+        /**
+         * @param landmark point3 in the world
```

Comment:
Say what it is :-)
brief constructor with known landmark, offset, and calibration.

Reply:
done